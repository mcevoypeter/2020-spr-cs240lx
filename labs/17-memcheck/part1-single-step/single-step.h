#ifndef __SINGLE_STEP_H__
#define __SINGLE_STEP_H__

// initialize single step mode.
int single_step_init(void);

// run a single routine <user_fn> in single step mode using stack <sp>
int single_step_run(int user_fn(void), uint32_t sp);

#endif
