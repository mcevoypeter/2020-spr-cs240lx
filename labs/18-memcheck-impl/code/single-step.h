#ifndef __SINGLE_STEP_H__
#define __SINGLE_STEP_H__

#include "rpi.h"
#include "cp14-debug.h"
#include "bit-support.h"

// initialize single step mode: 1 on success.
int single_step_init(void);

// run a single routine <user_fn> in single step mode using stack <sp>
int single_step_run(int user_fn(void), uint32_t sp);

// return the number of instructions we single stepped.
//  (just count the faults)
unsigned single_step_cnt(void);

#endif
