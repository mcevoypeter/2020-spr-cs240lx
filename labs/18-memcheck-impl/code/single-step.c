#include "rpi.h"
#include "rpi-interrupts.h"
#include "single-step.h"
#include "cp14-debug.h"
#include "libc/helper-macros.h"
#include "sys-call-asm.h"

int single_step_init(void) {
    cp14_enable();
    return 1;
}

static unsigned step_cnt = 0;
static void single_step_handler(uint32_t regs[16], uint32_t pc, uint32_t addr) {
    step_cnt++; 
    brkpt_mismatch_set0(pc, single_step_handler);    
}

int single_step_run(int user_fn(void), uint32_t sp) {
    step_cnt = 0;

    brkpt_mismatch_set0(0, single_step_handler);

    int ret_val = user_mode_run_fn(user_fn, sp);

    brkpt_mismatch_disable0(0);

    return ret_val;
}

unsigned single_step_cnt(void) {
    return step_cnt;
}
