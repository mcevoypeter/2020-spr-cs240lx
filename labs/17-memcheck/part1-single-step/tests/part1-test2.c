// engler: trivial check that you can switch to user mode and back.
#include "single-step.h"

static int user_fn(void) {
    trace("user mode = %b, current mode=%b\n", USER_MODE, mode_get(cpsr_get()));
    assert(mode_is_user());
    for(int i = 0; i < 10; i++) 
        trace("hello %d\n", i);
    return 4;
}

void notmain() {
    cp14_enable();

    assert(mode_is_super());
    trace("about to try running at user level\n");
    int x = user_mode_run_fn(user_fn, STACK_ADDR2);
    assert(x == 4);
    trace("success!\n");
}
