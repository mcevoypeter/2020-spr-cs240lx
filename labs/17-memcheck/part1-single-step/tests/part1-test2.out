find-ttyusb.c:find_ttyusb_first:106:FOUND: </dev/ttyUSB0>
opened tty port </dev/ttyUSB0>.
my-install: tty-usb=</dev/ttyUSB0> program=<./tests/part1-test2.bin> about to boot
******************sending 10092 bytes
waiting for a start
got a start
putting code
put code and got back
bootloader: Done.
listening on ttyusb=</dev/ttyUSB0>
TRACE:about to try running at user level
TRACE:user mode = 10000, current mode=10000
TRACE:hello 0
TRACE:hello 1
TRACE:hello 2
TRACE:hello 3
TRACE:hello 4
TRACE:hello 5
TRACE:hello 6
TRACE:hello 7
TRACE:hello 8
TRACE:hello 9
cur mode: SUPER MODE, saved mode = USER MODE, switch-to mode=SUPER MODE
shadow lr=0x8120, shadow sp = 0x7000000
pc = 0x8fcc, saved sp=0x7000000
system call: sp=0x7ffff90, inst=11101111000000000000000000000001, sys_num=1, r0=0x60000113
TRACE:success!
DONE!!!

Saw done

bootloader: pi exited.  cleaning up
